docs: update scale recommendation to account for read load (#921)
langchain-ai/langsmith-docs: docs/self_hosting/configuration/scale.mdx

# Configuring LangSmith for scale

> A self-hosted LangSmith instance can handle a large number of traces and users. The default configuration for the self-hosted deployment can handle substantial load, and you can configure your deployment to be able to achieve higher scale. Below, we walk through scaling considerations and provide some examples to help configure your self-hosted instance.

Reviewer: ```suggestion
A self-hosted LangSmith instance can handle a large number of traces and users. The default configuration for the self-hosted deployment can handle substantial load, and you can configure your deployment to be able to achieve higher scale. This page describes scaling considerations and provides some examples to help configure your self-hosted instance.
```

> For example configurations, check out our [scale examples](#example-langsmith-configurations-for-scale)

Reviewer: ```suggestion
For example configurations, refer to [Example LangSmith configurations for scale](#example-langsmith-configurations-for-scale).
```

## Trace ingestion (write path)

Services that play a large role in trace ingestion:

- Platform backend service: Receives initial request to ingest traces and places traces on a Redis queue
- Redis cache: Used to queue traces that need to be persisted
- Queue service: Persists traces for querying
- ClickHouse: Persistent storage used for traces

Common usage that put load on the write path:

- Ingesting traces via the Python or JavaScript LangSmith SDK
- Ingesting traces via the `@traceable` wrapper
- Submitting traces via the `/runs/multipart` endpoint

When scaling up the write path (trace ingestion), it is helpful to monitor the four services/resources listed above. Here are some typical changes that can help increase performance of trace ingestion:

- Give ClickHouse more resources (CPU and memory) if it is approaching resource limits.
- Increase the number of platform-backend pods if ingest requests are taking long to respond.
- Increase queue service pod replicas if traces are not being processed from Redis fast enough.

> - Use a larger Redis cache if you notice that the current Redis instance is reaching resource limits. This could also be a reason why ingest requests take a long time.

Reviewer: This is just a suggestion, so feel free to ignore as I'm just diving into the content/technical aspects here, but what do you think about switching the pattern around here a little. Mainly to keep the services and then the recommendations together, so that you don't have to refer back to them. Something like:

> ## Trace ingestion (write path)
> 
> Common usage that can put load on the write path include:
> 
> [... bullets points]
> 
> When scaling up the write path, it is helpful to monitor the service/resources that play a large role in trace ingestion. The following outlines the services and some recommendations that can help increase performance of trace ingestion:
> 
> - **Platform backend service** receives the initial request to ingest traces and places traces on a Redis queue. You can increase the number of platform-backend pods if ingest requests are taking a long time to respond.
> - **Redis cache** queues traces that need to be persisted. You can use a larger Redis cache if you notice that the existing Redis instance is reaching resource limits, which could also cause ingest requests to take a long time.
> - **Queue service** persists traces for querying. You can increase the queue service pod replicas if traces are not being processed from Redis fast enough.
> - **ClickHouse** is the persistent storage for traces. You can give ClickHouse more resource (CPU and memory) if it is approaching resource limits.

Author: I like keeping the service description with service configuration recommendations together. What Ill do here is Ill move the middle sections to the front so it goes: 1) usage for that path, 2) services that play a role, 3) how to update service configuration

instead of the current: 1) services that play a role, 2) usage for that path, 3) how to update service configuration

I could see combining the last two sections, but for now I like it split just slightly more bc the bullet points are pretty short. If we want to switch it later to combine the bottom two sections I wouldn't be too opposed

> ## Trace querying (read path)

Reviewer: If you go for the different structure pattern for that write path section, I would replicate that here!

Services that play a large role in querying traces:

- Backend service: Receives the request and submits a query to ClickHouse to then respond to the request
- ClickHouse: Persistent storage for traces. This is the main database that is queried when requesting trace info.

Common usage that puts load on the read path:

- Users on the frontend looking at tracing projects or individual traces
- Scripts used to query for trace info
- Hitting either the `/runs/query` or `/runs/<run-id>` api endpoints

When scaling up the read path (trace querying), it is helpful to monitor the two services/resources listed above. Here are some typical changes that can help improve performance of trace querying:

- Increase the number of backend service pods. This would be most impactful if backend service pods are reaching 1 core CPU usage.
- Give ClickHouse more resources (CPU or Memory). ClickHouse can be very resource intensive, but it should lead to better performance.
- Move to a [replicated ClickHouse cluster](./external_clickhouse#ha-replicated-clickhouse-cluster). Adding replicas of ClickHouse helps with read performance, but we recommend staying below 5 replicas (start with 3).

> For more precise guidance on how this translates to helm chart values, check out the examples below. If you are unsure why your LangSmith instance cannot handle a certain load pattern, please reach out to the LangChain team.

Reviewer: Linking to the section, just in case additional content or the page changes and we (I) forget to update the language. Also, do we have an email or link for the "contact the LangChain team" part?

```suggestion
For more precise guidance on how this translates to helm chart values, refer to the examples the following [section](#example-langsmith-configurations-for-scale). If you are unsure why your LangSmith instance cannot handle a certain load pattern, contact the LangChain team.
```

## Example LangSmith configurations for scale

Below we provide some example LangSmith configurations based on expected read and write loads.

For read load (trace querying):

- Low means roughly 5 users looking at traces at a time (about 10 requests per second)
- Medium means roughly 20 users looking at traces at a time (about 40 requests per second)
- High means roughly 50 users looking at traces at a time (about 100 requests per second)

For write load (trace ingestion):

- Low means up to 10 traces submitted per second
- Medium means up to 100 traces submitted per second
- High means up to 1000 traces submitted per second

:::important
The exact optimal configuration depends on your usage and trace payloads. Use the examples below in combination with the information above and your specific usage to update your LangSmith configuration as you see fit. If you have any questions, please reach out to the LangChain team.
:::

### Low reads, low writes

> Our default LangSmith configuration should be able to easily handle this load. No custom resource configuration is needed here.

Reviewer: ```suggestion
The default LangSmith configuration will manage this load. No custom resource configuration is needed here.
```
Does that sound ok?

### Low reads, high writes

> This means that you have a very high scale of trace ingestions, but single digit number of users on the frontend querying traces at any one time.

Reviewer: ```suggestion
You have a very high scale of trace ingestions, but single digit number of users on the frontend querying traces at any one time.
```

#   minReplicas: 8

## Note that we are actively working on improving performance of this service to reduce the number of replicas.
queue:
  deployment:
    replicas: 170 # OR enable autoscaling to this level (example below)
# autoscaling:
#   enabled: true
#   maxReplicas: 170
#   minReplicas: 40

backend:
  deployment:
    replicas: 5 # OR enable autoscaling to this level (example below)
# autoscaling:
#   enabled: true
#   maxReplicas: 5
#   minReplicas: 3

## Ensure your Redis cache is at least 200 GB
redis:
  external:
    enabled: true
    existingSecretName: langsmith-redis-secret # Set the connection url for your external Redis instance (200+ GB)

clickhouse:
  statefulSet:
    persistence:
      # This may depend on your configured TTL (see config section).
      # We recommend 600Gi for every shortlived TTL day if operating at this scale constantly.
      size: 4200Gi # This assumes 7 days TTL and operating a this scale constantly.
    resources:
      requests:
        cpu: "10"
        memory: "32Gi"
      limits:
        cpu: "16"
        memory: "48Gi"

commonEnv:
  - name: "CLICKHOUSE_ASYNC_INSERT_WAIT_PCT_FLOAT"
    value: "0"
```

### High reads, low writes

> This means that you have a relatively low scale of trace ingestions, but many frontend users querying traces and/or have scripts that hit the `/runs/query` or `/runs/<run-id>` endpoints frequently.

Reviewer: ```suggestion
You have a relatively low scale of trace ingestions, but many frontend users querying traces and/or have scripts that hit the `/runs/query` or `/runs/<run-id>` endpoints frequently.
```
